tunnel-server/src/proxy.ts

```diff
@@ -54,17 +69,33 @@ export function proxyHandlers({
         return;
       }
 
-      if(env.access === 'private'){
-        const authenticate = authenticator([tunnelingKeyAuthenticator(env.publicKey)])
-        try {
-          const claims = await authenticate(req)  
-          if (!claims){
-            return unauthorized(res);
+      const session = sessionManager(req, res, env.publicKeyThumbprint)
+      if (env.access === "private"){
+        if (!session.user){
+          const authenticate = authenticator([JwtAuthenticator(env)])
+          try {
+            const authResult = await authenticate(req)
+            if (!authResult.isAuthenticated){
+              return unauthorized(res);
+            }
+            session.set(authResult.claims)
+            if (authResult.login && req.method === 'GET'){
+              session.save()
+              redirectToLogin(res, env.hostname, req.url)
+              return;
+            } 
+            if (authResult.method.type === 'header'){
```

Comment:
this is a smelly side-effect, how about having the authenticator transform the headers?